From 4cd231fa58cd075f86bd1abc4c17f63f09677d53 Mon Sep 17 00:00:00 2001
From: Yann Gautier <yann.gautier@foss.st.com>
Date: Fri, 13 Aug 2021 14:17:01 +0200
Subject: [PATCH 152/954] feat(stm32mp2): pen holding for core 1

After boot, the BootROM will issue a reset of the Cortex-A35 cores.
Core 0 will boot till kernel. Core 1 is placed in WFI in TF-A BL2.
Through PSCI, Linux will ask to have Core 1 available. The PSCI platform
code is done in stm32_pwr_domain_on(). Core0 will change Core1 reset
address thanks to CA35SS_SYSCFG_VBAR_CR register to BL31 entry point.
And will reset Core1 by setting RCC_C1P1RSTCSETR_C1P1PORRST bit.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 93c1527c3016a04e5ba97e79317baec65c5cedf9 feat(stm32mp2): pen holding for core 1]
Signed-off-by: Yann Gautier <yann.gautier@foss.st.com>
Change-Id: I01e59e3a2398c48cc050ec4703d6610da9e9c4bd
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/234835
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
---
 plat/st/stm32mp2/stm32mp2_pm.c | 38 +++++++++++++++++++++++++++++++++-
 1 file changed, 37 insertions(+), 1 deletion(-)

diff --git a/plat/st/stm32mp2/stm32mp2_pm.c b/plat/st/stm32mp2/stm32mp2_pm.c
index 869c449d4b..4ebf5dc632 100644
--- a/plat/st/stm32mp2/stm32mp2_pm.c
+++ b/plat/st/stm32mp2/stm32mp2_pm.c
@@ -18,15 +18,42 @@
 
 #include <platform_def.h>
 
+#define CA35SS_SYSCFG_VBAR_CR	0x2084U
+
 static uintptr_t stm32_sec_entrypoint;
+static u_register_t cntfrq_core0;
 
 static void stm32_cpu_standby(plat_local_state_t cpu_state)
 {
 }
 
+/*******************************************************************************
+ * STM32MP2 handler called when a power domain is about to be turned on. The
+ * mpidr determines the CPU to be turned on.
+ * Called by core 0 to activate core 1.
+ ******************************************************************************/
 static int stm32_pwr_domain_on(u_register_t mpidr)
 {
-	return PSCI_E_INTERN_FAIL;
+	unsigned long current_cpu_mpidr = read_mpidr();
+
+	if (stm32mp_is_single_core()) {
+		return PSCI_E_INTERN_FAIL;
+	}
+
+	if (mpidr == current_cpu_mpidr) {
+		return PSCI_E_INVALID_PARAMS;
+	}
+
+	cntfrq_core0 = read_cntfrq_el0();
+	flush_dcache_range((uintptr_t)&cntfrq_core0, sizeof(u_register_t));
+
+	/* Set CA35SS_SYSCFG VBAR address to secure entrypoint */
+	mmio_write_32(A35SSC_BASE + CA35SS_SYSCFG_VBAR_CR, stm32_sec_entrypoint);
+
+	/* Reset CPU1 */
+	mmio_write_32(RCC_BASE + RCC_C1P1RSTCSETR, RCC_C1P1RSTCSETR_C1P1PORRST);
+
+	return PSCI_E_SUCCESS;
 }
 
 static void stm32_pwr_domain_off(const psci_power_state_t *target_state)
@@ -39,8 +66,17 @@ static void stm32_pwr_domain_suspend(const psci_power_state_t *target_state)
 	/* Nothing to do, power domain is not disabled */
 }
 
+/*******************************************************************************
+ * STM32MP2 handler called when a power domain has just been powered on after
+ * being turned off earlier. The target_state encodes the low power state that
+ * each level has woken up from.
+ * Called by core 1 just after wake up.
+ ******************************************************************************/
 static void stm32_pwr_domain_on_finish(const psci_power_state_t *target_state)
 {
+	stm32mp_gic_pcpu_init();
+
+	write_cntfrq_el0(cntfrq_core0);
 }
 
 /*******************************************************************************
-- 
2.53.0

